File: 1005.maximize-sum-of-array-after-k-negations.cpp
/*
 * @lc app=leetcode id=1005 lang=cpp
 *
 * [1005] Maximize Sum Of Array After K Negations
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int largestSumAfterKNegations(vector<int>& nums, int k) {
    int neg_num = 0;
    for (const auto& num : nums) {
      if (num < 0) {
        neg_num++;
      }
    }
    if (k >= neg_num) {
      for (auto& num : nums) {
        if (num < 0) {
          num = -1 * num;
        }
      }
      k -= neg_num;
      k %= 2;
      if (k == 0) {
        return accumulate(nums.begin(), nums.end(), 0);
      }

      int min_num = INT_MAX;
      for (const auto& num : nums) {
        min_num = min(min_num, num);
      }
      return accumulate(nums.begin(), nums.end(), -2 * min_num);
    }
    sort(nums.begin(), nums.end());
    for (int i = 0; i < k; i++) {
      nums[i] = -1 * nums[i];
    }
    return accumulate(nums.begin(), nums.end(), 0);
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> nums;
  nums = {3, -1, 0, 2};
  ASSERT_EQ(s.largestSumAfterKNegations(nums, 3), 6);
  nums = {4, 2, 3};
  ASSERT_EQ(s.largestSumAfterKNegations(nums, 1), 5);
  nums = {2, -3, -1, 5, -4};
  ASSERT_EQ(s.largestSumAfterKNegations(nums, 2), 13);
  return 0;
}

File: 1005.maximize-sum-of-array-after-k-negations.rs
/*
 * @lc app=leetcode id=1005 lang=rust
 *
 * [1005] Maximize Sum Of Array After K Negations
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start

use std::cmp::Reverse;
use std::collections::BinaryHeap;
impl Solution {
    pub fn largest_sum_after_k_negations(nums: Vec<i32>, mut k: i32) -> i32 {
        let mut pq = BinaryHeap::new();
        for num in &nums {
            pq.push(Reverse(*num));
        }
        while k > 0 {
            if let Some(Reverse(n)) = pq.pop() {
                pq.push(Reverse(-n));
            }
            k -= 1;
        }
        let mut sum = 0;
        for _ in 0..nums.len() {
            if let Some(Reverse(n)) = pq.pop() {
                sum += n
            }
        }
        sum
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::largest_sum_after_k_negations(vec![4, 2, 3], 1), 5);
}

#[test]
fn test2() {
    assert_eq!(
        Solution::largest_sum_after_k_negations(vec![3, -1, 0, 2], 3),
        6
    );
}

#[test]
fn test3() {
    assert_eq!(
        Solution::largest_sum_after_k_negations(vec![2, -3, -1, 5, -4], 2),
        13
    );
}
